test/integration/replicaset/replicaset_test.go

```diff
@@ -272,246 +280,595 @@ func TestAdoption(t *testing.T) {
 	defer framework.DeleteTestingNamespace(ns, s, t)
-	rs := newRS("rs", ns.Name, 2)
-	pod1 := newMatchingPod("pod1", ns.Name)
-	pod2 := newMatchingPod("pod2", ns.Name)
-	createRSsPods(t, clientSet, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{pod1, pod2}, ns.Name)
 
 	stopCh := make(chan struct{})
 	informers.Start(stopCh)
 	go rm.Run(5, stopCh)
-	waitRSStable(t, clientSet, rs, ns.Name)
+	defer close(stopCh)
 
-	// change the rs's selector to match both pods
-	patch := `{"metadata":{"labels":{"name":null}}}`
-	podClient := clientSet.Core().Pods(ns.Name)
-	pod2, err := podClient.Patch(pod2.Name, types.StrategicMergePatchType, []byte(patch))
+	labelMap := map[string]string{"foo": "bar"}
+
+	// Create 2 RS with identical selectors
+	for i := 0; i < 2; i++ {
+		// One RS has 1 replica, and another has 2 replicas
+		rs := newRS(fmt.Sprintf("rs-%d", i+1), ns.Name, i+1)
+		rs.Spec.Selector.MatchLabels = labelMap
+		rs.Spec.Template.Labels = labelMap
+		rss, _ := createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{}, ns.Name)
+		waitRSStable(t, c, rss[0], ns.Name)
+	}
+
+	// Expect 3 total Pods to be created
+	podSelector := labels.Set{"foo": "bar"}.AsSelector()
+	options := metav1.ListOptions{LabelSelector: podSelector.String()}
+	pods, err := c.Core().Pods(ns.Name).List(options)
 	if err != nil {
-		t.Fatalf("Failed to patch pod2: %v", err)
+		t.Fatalf("Failed obtaining a list of pods that match the pod labels %v: %v", labelMap, err)
+	}
+	if pods == nil {
+		t.Fatalf("Obtained a nil list of pods")
+	}
+	if len(pods.Items) != 3 {
+		t.Errorf("len(pods) = %v, want %v", len(pods.Items), 3)
+	}
+
+	// Expect both RSs have status.replicas = spec.replicas
+	for i := 0; i < 2; i++ {
+		newRS, err := c.ExtensionsV1beta1().ReplicaSets(ns.Name).Get(fmt.Sprintf("rs-%d", i+1), metav1.GetOptions{})
+		if err != nil {
+			t.Fatalf("failed to get replicaset rs-%d: %v", i+1, err)
+		}
+		if newRS.Status.Replicas != *newRS.Spec.Replicas {
+			t.Fatalf(".Status.Replicas %d is not equal to .Spec.Replicas %d", newRS.Status.Replicas, *newRS.Spec.Replicas)
+		}
+	}
+}
+
+func TestPodOrphaningAndAdoptionWhenLabelsChange(t *testing.T) {
+	s, closeFn, rm, informers, c := rmSetup(t)
+	defer closeFn()
+	ns := framework.CreateTestingNamespace("test-pod-orphaning-and-adoption-when-labels-change", s, t)
+	defer framework.DeleteTestingNamespace(ns, s, t)
+	labelMap := map[string]string{"foo": "bar"}
+	rs := newRS("rs", ns.Name, 1)
+	rs.Spec.Selector.MatchLabels = labelMap
+	rs.Spec.Template.Labels = labelMap
+	rss, _ := createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{}, ns.Name)
+	rs = rss[0]
+
+	stopCh := make(chan struct{})
+	informers.Start(stopCh)
+	go rm.Run(5, stopCh)
+	defer close(stopCh)
+	waitRSStable(t, c, rs, ns.Name)
+
+	// Orphaning: RS should remove OwnerReference from a pod when the pod's labels change to not match its labels
+	// Although we can access the pod directly, we should still access it via selector to ensure the selector is working
```

Comment:
This line is no longer applicable. You have to use the selector to find the Pod that RS created.

Reply:
It looks to me that both the comment and your clarification support usage of the selector, but the reason given by the comment is wrong (we are not testing that the selector is working). I will rephrase the comment accordingly.